File: apps/android/app/src/test/kotlin/com/loonext/android/core/model/EmergencyFlagTest.kt
package com.loonext.android.core.model

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * #414 / #565 — the urgent mark, and that this phone agrees with the laptop and
 * with iOS about when it shows.
 *
 * Two halves. The behaviour tests assert the rule; the parity tests read
 * `packages/shared/src/emergency-flag.ts`, because this is a hand-port and
 * nothing about Kotlin says the original stayed put.
 *
 * The bug this file exists for is one layer up: `ConversationDetail` never
 * declared `emergency_at`, so `ignoreUnknownKeys` dropped it and the thread an
 * urgent notification opens was the one screen that could not say why you were
 * there. `scripts/check-conversation-detail-parity.mjs` guards that half — a
 * model missing a field the server sends. This file guards the rule that reads it.
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class EmergencyFlagTest {

    private val WHEN = "2026-08-08T23:04:00.000Z"

    @Test
    fun `an ordinary open thread is not flagged`() {
        assertFalse(isConversationFlaggedUrgent(null, null))
    }

    @Test
    fun `an urgent thread is flagged while it is open`() {
        assertTrue(isConversationFlaggedUrgent(WHEN, null))
    }

    @Test
    fun `closing the thread clears the mark`() {
        // Closing is the product's existing word for "handled". A badge that
        // never cleared would be decoration, and a timer deciding an emergency
        // stopped mattering would be a guess made while somebody was still
        // driving to it.
        assertFalse(isConversationFlaggedUrgent(WHEN, WHEN))
    }

    @Test
    fun `a closed thread that was never urgent is not flagged`() {
        assertFalse(isConversationFlaggedUrgent(null, WHEN))
    }

    @Test
    fun `the label is not pre-shouted, so TalkBack does not spell it`() {
        assertEquals("Urgent", URGENT_BADGE_LABEL)
        assertTrue(URGENT_BADGE_LABEL != URGENT_BADGE_LABEL.uppercase())
    }

    // --- Against the original ------------------------------------------------

    private fun repoFile(relative: String): File {
        var dir = File("").absoluteFile
        while (true) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate
            dir = dir.parentFile ?: break
        }
        throw AssertionError("$relative is not reachable from ${File("").absolutePath}")
    }

    private fun sharedSource(): String =
        repoFile("packages/shared/src/emergency-flag.ts").readText()

    @Test
    fun `the shared module still reads presence rather than an ordering`() {
        // A grep rather than a second implementation: the one way this rule could
        // change without any behaviour test here failing is if the original
        // started COMPARING the two timestamps. Both answers are "a boolean", so
        // only the source shows the difference.
        val shared = sharedSource()
        assertTrue(
            "the shared rule is no longer a presence check",
            shared.contains(
                "conversation.emergency_at !== null && conversation.closed_at === null",
            ),
        )
    }

    @Test
    fun `the shared label is the word this file expects`() {
        val shared = sharedSource()
        assertTrue(
            "URGENT_BADGE_LABEL has changed in the shared module",
            shared.contains("URGENT_BADGE_LABEL = \"$URGENT_BADGE_LABEL\""),
        )
    }
}

File: packages/shared/src/emergency-flag.test.ts
import { describe, expect, it } from "vitest";

import {
  isFlaggedUrgent,
  URGENT_BADGE_LABEL,
  type EmergencyFlagFields,
} from "./emergency-flag";

const thread = (
  fields: Partial<EmergencyFlagFields> = {},
): EmergencyFlagFields => ({
  emergency_at: null,
  closed_at: null,
  ...fields,
});

const WHEN = "2026-08-08T23:04:00.000Z";

describe("isFlaggedUrgent (#414 / #565)", () => {
  it("is false for an ordinary open thread", () => {
    expect(isFlaggedUrgent(thread())).toBe(false);
  });

  it("is true while an urgent thread is still open", () => {
    expect(isFlaggedUrgent(thread({ emergency_at: WHEN }))).toBe(true);
  });

  it("clears when the crew closes the thread", () => {
    // Closing is the product's existing word for "handled". A badge that never
    // cleared would be decoration, and a timer deciding an emergency stopped
    // mattering would be a guess made while somebody was still driving to it.
    expect(
      isFlaggedUrgent(thread({ emergency_at: WHEN, closed_at: WHEN })),
    ).toBe(false);
  });

  it("does not fire on a closed thread that was never urgent", () => {
    expect(isFlaggedUrgent(thread({ closed_at: WHEN }))).toBe(false);
  });

  it("reads presence, not an ordering — and does not need to", () => {
    /**
     * An urgent reply arriving after a close looks like it should matter, and it
     * cannot happen. The inbound that sets `emergency_at` goes through
     * `messages_automated`, which REOPENS a thread closed within 30 days
     * (`set status = 'new', closed_at = null`) and otherwise starts a fresh row.
     * So a row never carries an `emergency_at` later than its own `closed_at`.
     *
     * Asserted anyway, and asserted as PRESENCE, because comparing the two
     * timestamps would be a second rule — one guarding a state the database
     * cannot produce, drifting quietly if that reopen window ever changes.
     */
    expect(
      isFlaggedUrgent(
        thread({
          emergency_at: "2026-08-08T23:04:00.000Z",
          closed_at: "2026-08-01T09:00:00.000Z",
        }),
      ),
    ).toBe(false);
  });

  it("says Urgent, so a screen reader does not spell it", () => {
    // The clients upper-case it in their own styling. If this constant were
    // already upper-case, VoiceOver would read it letter by letter.
    expect(URGENT_BADGE_LABEL).toBe("Urgent");
    expect(URGENT_BADGE_LABEL).not.toBe(URGENT_BADGE_LABEL.toUpperCase());
  });
});
